File: programmers/algorithm_data_structure/graph/가장 먼 노드.py
from collections import defaultdict


def my_solution(n, edge):
    '''
    graph 구조는 dictionay(hash) 구조 사용
    - defaultdict로 코드 크기도 줄이고
    - (딕셔너리 컴프리헨션에 비해) 속도가 빠르다.

    매번 방문하지 않은 인접한 노드를 찾아 탐색해가면
    모든 노드를 최단 거리로 만난다.

    마지막 노드를 방문하기 전에 루프문을 나와서
    마지막으로 탐색할 노드 개수를 답으로 반환
    '''
    # graph 생성
    graph = defaultdict(set)
    for n1, n2 in edge:
        graph[n1].add(n2)
        graph[n2].add(n1)

    # 반복: 방문하지 않은 노드 탐색
    visited = set()
    next_visit = set([1])
    while True:
        # 방문한 노드 업데이트
        visited.update(next_visit)

        # 마지막 방문할 노드 차례인 경우 분기 종료
        if len(visited) == n:
            break

        # 다음 방문할 노드 업데이트
        temp = set()
        for node in next_visit:
            temp.update([n for n in graph[node] if n not in visited])
        next_visit = temp

    return len(next_visit)


def worse_solution(n, edge):
    # graph 생성
    graph = defaultdict(set)
    for n1, n2 in edge:
        graph[n1].add(n2)
        graph[n2].add(n1)

    # 반복: 방문하지 않은 노드 탐색
    visited = [0] * n
    next_visit = set([1])
    while True:
        # 방문한 노드 업데이트
        for node in next_visit:
            visited[node-1] = 1

        # 마지막 방문할 노드 차례인 경우 분기 종료
        if sum(visited) == n:
            break

        # 다음 방문할 노드 업데이트
        temp = set()
        for node in next_visit:
            temp.update([gn for gn in graph[node] if visited[gn-1] == 0])
        next_visit = temp

    return len(next_visit)

File: programmers/algorithm_data_structure/graph/가장_먼_노드.java
package algorithm_data_structure.graph;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class 가장_먼_노드 {
    /**
     * @param n     : 노드의 갯수(1번부터 n번까지 존재)
     * @param edge  : 노드 간선 집합, (a, b)는 a와 b를 잇는 간선임
     * @return      : 1번 노드에서 가장 멀리 떨어진 노드의 갯수
     *                (각 노드로 가는 최단거리로 거리 구분)
     * 
     * map, set 자료구조를 활용해 그래프를 만들고
     * 매번 방문하지 않은 노드를 찾아 나서며 최단 거리 탐색
     * (사이클 방지)
     */
    public int mySolution(int n, int[][] edge) {
        int answer = 0; 
        boolean[] visited = new boolean[n+1];
        Map<Integer, Set<Integer>> graph = getGraph(n, edge);

        List<Integer> currNodes = new ArrayList<>();
        currNodes.add(1);
        visited[1] = true;

        while (!currNodes.isEmpty()) {
            List<Integer> nextNodes = new ArrayList<>();

            for (int node : currNodes) {
                for (int next : graph.get(node)) {
                    if (!visited[next]) {
                        nextNodes.add(next);
                        visited[next] = true;
                    }
                }
            }
            
            answer = currNodes.size();
            currNodes = nextNodes;
        }

        return answer;
    }    
    
    public Map<Integer, Set<Integer>> getGraph(int n, int[][] edge) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();

        for (int i = 1; i <= n; i++) {
            graph.put(i, new HashSet<>());
        }

        for (int[] e : edge) {
            graph.get(e[0]).add(e[1]);
            graph.get(e[1]).add(e[0]);
        }

        return graph;
    }

    public int mySolution2(int n, int[][] edge) {
        int answer = 0;

        return answer;
    }
}